File: backend/services/koreanStock.js
import yahooFinance from 'yahoo-finance2';
import axios from 'axios';

const DART_API_KEY = process.env.DART_API_KEY || '';

function fmtKrw(val) {
    if (val == null || isNaN(val)) return 'N/A';
    return `${Math.round(val).toLocaleString('ko-KR')}원`;
}

function fmtLargeKrw(val) {
    if (val == null || isNaN(val)) return 'N/A';
    if (val >= 1e12) return `${(val / 1e12).toFixed(0)}조 원`;
    if (val >= 1e8) return `${(val / 1e8).toFixed(0)}억 원`;
    return `${val.toLocaleString('ko-KR')}원`;
}

function safe(val, fallback = 'N/A') {
    if (val == null || (typeof val === 'number' && isNaN(val))) return fallback;
    return val;
}

async function getDartCorpCode(ticker) {
    if (!DART_API_KEY) return null;
    try {
        const res = await axios.get('https://opendart.fss.or.kr/api/company.json', {
            params: { crtfc_key: DART_API_KEY, stock_code: ticker },
            timeout: 5000,
        });
        return res.data?.corp_code || null;
    } catch {
        return null;
    }
}

async function getDartFinancials(ticker) {
    if (!DART_API_KEY) return [];
    const corpCode = await getDartCorpCode(ticker);
    if (!corpCode) return [];

    const currentYear = new Date().getFullYear();
    const rows = [];

    for (let year = currentYear - 3; year < currentYear; year++) {
        try {
            const res = await axios.get('https://opendart.fss.or.kr/api/fnlttSinglAcntAll.json', {
                params: {
                    crtfc_key: DART_API_KEY,
                    corp_code: corpCode,
                    bsns_year: year,
                    reprt_code: '11011', // 사업보고서
                    fs_div: 'CFS',       // 연결재무제표
                },
                timeout: 8000,
            });

            const list = res.data?.list || [];
            const extract = (...keywords) => {
                const item = list.find(r =>
                    keywords.some(k => r.account_nm?.includes(k)) && r.sj_div === 'IS'
                );
                if (!item) return 0;
                return parseInt((item.thstrm_amount || '0').replace(/,/g, ''), 10) || 0;
            };

            // 연결 재무제표가 없으면 별도 재무제표로 재시도
            let rev = extract('매출액', '수익(매출액)');
            let op  = extract('영업이익');
            let net = extract('당기순이익');

            if (!rev) {
                const res2 = await axios.get('https://opendart.fss.or.kr/api/fnlttSinglAcntAll.json', {
                    params: {
                        crtfc_key: DART_API_KEY,
                        corp_code: corpCode,
                        bsns_year: year,
                        reprt_code: '11011',
                        fs_div: 'OFS', // 별도재무제표
                    },
                    timeout: 8000,
                });
                const list2 = res2.data?.list || [];
                const extract2 = (...keywords) => {
                    const item = list2.find(r =>
                        keywords.some(k => r.account_nm?.includes(k)) && r.sj_div === 'IS'
                    );
                    if (!item) return 0;
                    return parseInt((item.thstrm_amount || '0').replace(/,/g, ''), 10) || 0;
                };
                rev = extract2('매출액', '수익(매출액)');
                op  = extract2('영업이익');
                net = extract2('당기순이익');
            }

            // DART 금액 단위는 원 → 억 원으로 변환
            rows.push({ year: String(year), rev: Math.round(rev / 1e8), op: Math.round(op / 1e8), net: Math.round(net / 1e8) });
        } catch {
            // 해당 연도 데이터 없으면 스킵
        }
    }
    return rows;
}

export async function getKoreanStock(ticker) {
    const yfTicker = `${ticker}.KS`;

    let quote, summary;
    try {
        [quote, summary] = await Promise.all([
            yahooFinance.quote(yfTicker),
            yahooFinance.quoteSummary(yfTicker, {
                modules: ['defaultKeyStatistics', 'summaryDetail', 'financialData', 'assetProfile'],
            }),
        ]);
    } catch {
        // KOSDAQ 시도
        try {
            const yfTickerKQ = `${ticker}.KQ`;
            [quote, summary] = await Promise.all([
                yahooFinance.quote(yfTickerKQ),
                yahooFinance.quoteSummary(yfTickerKQ, {
                    modules: ['defaultKeyStatistics', 'summaryDetail', 'financialData', 'assetProfile'],
                }),
            ]);
        } catch (e) {
            return { error: `'${ticker}' 종목 데이터를 찾을 수 없습니다.` };
        }
    }

    const ks = summary?.defaultKeyStatistics || {};
    const sd = summary?.summaryDetail || {};
    const fd = summary?.financialData || {};
    const ap = summary?.assetProfile || {};

    const price      = safe(quote?.regularMarketPrice);
    const prevClose  = safe(quote?.regularMarketPreviousClose);
    const changeVal  = price !== 'N/A' && prevClose !== 'N/A' ? price - prevClose : null;
    const changePct  = changeVal != null && prevClose ? (changeVal / prevClose) * 100 : null;

    const per  = safe(sd?.trailingPE);
    const pbr  = safe(ks?.priceToBook);
    const eps  = safe(ks?.trailingEps);
    const mktcap = safe(quote?.marketCap);
    const roe  = safe(fd?.returnOnEquity);
    const debt = safe(fd?.debtToEquity);
    const evebitda = safe(ks?.enterpriseToEbitda);
    const div  = safe(sd?.dividendYield);
    const beta = safe(ks?.beta);
    const vol  = safe(quote?.regularMarketVolume, 0);
    const avgvol = safe(sd?.averageVolume, 0);
    const high52 = safe(sd?.fiftyTwoWeekHigh);
    const low52  = safe(sd?.fiftyTwoWeekLow);

    const income = await getDartFinancials(ticker);

    const now = new Date();
    const asof = `${now.getFullYear()}.${String(now.getMonth()+1).padStart(2,'0')}.${String(now.getDate()).padStart(2,'0')} 종가 기준`;

    return {
        ticker,
        name:    safe(quote?.longName || quote?.shortName, ticker),
        sector:  safe(ap?.sector || ap?.industry, 'N/A'),
        price:   price !== 'N/A' ? fmtKrw(price) : 'N/A',
        change:  changeVal != null
            ? `${changeVal >= 0 ? '+' : ''}${fmtKrw(changeVal)} (${changePct >= 0 ? '+' : ''}${changePct.toFixed(2)}%)`
            : 'N/A',
        pos:     changeVal != null ? changeVal >= 0 : true,
        asof,
        per:      per !== 'N/A' ? `${Number(per).toFixed(1)}x` : 'N/A',
        pbr:      pbr !== 'N/A' ? `${Number(pbr).toFixed(1)}x` : 'N/A',
        eps:      eps !== 'N/A' ? fmtKrw(eps) : 'N/A',
        mktcap:   mktcap !== 'N/A' ? fmtLargeKrw(mktcap) : 'N/A',
        roe:      roe !== 'N/A' ? `${(roe * 100).toFixed(1)}%` : 'N/A',
        debt:     debt !== 'N/A' ? `${Number(debt).toFixed(1)}%` : 'N/A',
        evebitda: evebitda !== 'N/A' ? `${Number(evebitda).toFixed(1)}x` : 'N/A',
        div:      div !== 'N/A' ? `${(div * 100).toFixed(2)}%` : 'N/A',
        beta:     beta !== 'N/A' ? `${Number(beta).toFixed(2)}` : 'N/A',
        foreign:  'N/A',
        vol:      String(vol),
        avgvol:   String(avgvol),
        high52:   high52 !== 'N/A' ? fmtKrw(high52) : 'N/A',
        low52:    low52  !== 'N/A' ? fmtKrw(low52)  : 'N/A',
        income,
    };
}

File: backend/services/korean_stock.py
from datetime import datetime, timedelta
from concurrent.futures import ThreadPoolExecutor, as_completed
import time
from pykrx import stock as pykrx_stock
import FinanceDataReader as fdr
try:
    import OpenDartReader as odr
except ImportError:
    from opendartreader import OpenDartReader as odr
import yfinance as yf
import requests
from bs4 import BeautifulSoup
import os

DART_API_KEY = os.getenv("DART_API_KEY", "")
_NAVER_HEADERS = {"User-Agent": "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 Chrome/120"}

# 모듈 레벨 캐시 (서버 재시작 전까지 유지)
_krx_cache = {"data": None, "ts": 0}
_dart_cache = {"dart": None, "codes": None, "ts": 0}
_KRX_TTL  = 6 * 3600   # 6시간
_DART_TTL = 24 * 3600  # 24시간


def _today() -> str:
    return datetime.now().strftime("%Y%m%d")


def _date_n_days_ago(n: int) -> str:
    return (datetime.now() - timedelta(days=n)).strftime("%Y%m%d")


def _fmt_krw(value: float) -> str:
    return f"{int(value):,}원"


def _fmt_large_krw(value: float) -> str:
    if value >= 1_000_000_000_000:
        return f"{value / 1_000_000_000_000:.0f}조 원"
    if value >= 100_000_000:
        return f"{value / 100_000_000:.0f}억 원"
    return _fmt_krw(value)


def _safe(val, fallback="N/A"):
    if val is None or (isinstance(val, float) and val != val):
        return fallback
    return val


def _get_krx_listing():
    now = time.time()
    if _krx_cache["data"] is None or now - _krx_cache["ts"] > _KRX_TTL:
        _krx_cache["data"] = fdr.StockListing("KRX")
        _krx_cache["ts"] = now
    return _krx_cache["data"]


def _get_dart_instance():
    now = time.time()
    if _dart_cache["codes"] is None or now - _dart_cache["ts"] > _DART_TTL:
        dart = odr(DART_API_KEY)
        _dart_cache["dart"] = dart
        _dart_cache["codes"] = dart.corp_codes
        _dart_cache["ts"] = now
    return _dart_cache["dart"], _dart_cache["codes"]


def _get_name_sector_mktcap(ticker: str) -> tuple[str, str, str]:
    try:
        listing = _get_krx_listing()
        row = listing[listing["Code"] == ticker]
        if not row.empty:
            name   = str(row.iloc[0].get("Name", "") or "")
            sector = str(row.iloc[0].get("Industry", "") or row.iloc[0].get("Sector", "") or "")
            mc     = float(row.iloc[0].get("Marcap", 0) or 0)
            mktcap = _fmt_large_krw(mc) if mc else "N/A"
            return name, sector, mktcap
    except Exception:
        pass
    return ticker, "", "N/A"


def _get_naver_fundamentals(ticker: str) -> dict:
    try:
        url = f"https://finance.naver.com/item/main.naver?code={ticker}"
        r = requests.get(url, headers=_NAVER_HEADERS, timeout=8)
        soup = BeautifulSoup(r.text, "html.parser")

        def em(id_: str) -> str:
            tag = soup.find("em", id=id_)
            return tag.text.strip().replace(",", "") if tag else ""

        per = em("_per")
        eps = em("_eps")
        pbr = em("_pbr")
        dvr = em("_dvr")
        return {
            "per": f"{float(per):.1f}x" if per else "N/A",
            "pbr": f"{float(pbr):.1f}x" if pbr else "N/A",
            "eps": f"{int(eps):,}원"    if eps else "N/A",
            "div": f"{float(dvr):.2f}%" if dvr else "N/A",
        }
    except Exception:
        return {}


def _get_yf_extra(ticker: str) -> dict:
    for suffix in [".KS", ".KQ"]:
        try:
            info = yf.Ticker(ticker + suffix).info
            if info and info.get("returnOnEquity") is not None:
                roe  = _safe(info.get("returnOnEquity"))
                debt = _safe(info.get("debtToEquity"))
                ev   = _safe(info.get("enterpriseToEbitda"))
                beta = _safe(info.get("beta"))
                return {
                    "roe":      f"{roe * 100:.1f}%"  if roe  != "N/A" else "N/A",
                    "debt":     f"{debt:.1f}%"        if debt != "N/A" else "N/A",
                    "evebitda": f"{ev:.1f}x"          if ev   != "N/A" else "N/A",
                    "beta":     f"{beta:.2f}"          if beta != "N/A" else "N/A",
                }
        except Exception:
            continue
    return {"roe": "N/A", "debt": "N/A", "evebitda": "N/A", "beta": "N/A"}


def _fetch_finstate_year(dart, corp_code: str, year: int) -> dict | None:
    try:
        fs = dart.finstate(corp_code, year, "11011")
        if fs is None or fs.empty:
            fs = dart.finstate(corp_code, year, "11001")
        if fs is None or fs.empty:
            return None
        is_df = fs[fs["sj_div"] == "IS"]

        def extract(keywords):
            for kw in keywords:
                m = is_df[is_df["account_nm"].str.contains(kw, na=False)]
                if not m.empty:
                    raw = str(m.iloc[0].get("thstrm_amount", "0") or "0")
                    try:
                        return int(raw.replace(",", ""))
                    except Exception:
                        return 0
            return 0

        rev = extract(["매출액", "수익(매출액)"])
        op  = extract(["영업이익"])
        net = extract(["당기순이익"])
        return {"year": str(year), "rev": rev // 100_000_000, "op": op // 100_000_000, "net": net // 100_000_000}
    except Exception:
        return None


def get_dart_financials(ticker: str) -> list[dict]:
    if not DART_API_KEY:
        return []
    try:
        dart, codes = _get_dart_instance()
        match = codes[codes["stock_code"] == ticker]
        if match.empty:
            return []
        corp_code = match.iloc[0]["corp_code"]
        current_year = datetime.now().year
        years = range(current_year - 3, current_year)

        # 3개 연도 병렬 조회
        results = {}
        with ThreadPoolExecutor(max_workers=3) as ex:
            futs = {ex.submit(_fetch_finstate_year, dart, corp_code, y): y for y in years}
            for f in as_completed(futs):
                y = futs[f]
                row = f.result()
                if row:
                    results[y] = row
        return [results[y] for y in sorted(results)]
    except Exception:
        return []


def _get_naver_income(ticker: str) -> list[dict]:
    try:
        url = f"https://m.stock.naver.com/api/stock/{ticker}/finance/summary"
        r = requests.get(url, headers=_NAVER_HEADERS, timeout=8)
        data = r.json()
        annual = data.get("chartIncomeStatement", {}).get("annual", {})
        cols = annual.get("columns", [])
        if not cols or len(cols) < 3:
            return []
        years = cols[0][1:]
        revs  = cols[1][1:]
        ops   = cols[2][1:]
        rows = []
        for i, year_label in enumerate(years[:3]):
            year = year_label.replace(".", "")[:4]
            try:
                rev = int(str(revs[i]).replace(",", "")) if revs[i] else 0
                op  = int(str(ops[i]).replace(",", ""))  if ops[i]  else 0
            except Exception:
                rev, op = 0, 0
            rows.append({"year": year, "rev": rev, "op": op, "net": 0})
        return rows
    except Exception:
        return []


def _get_foreign(ticker: str, from_1m: str, today: str) -> str:
    try:
        f_df = pykrx_stock.get_exhaustion_rates_of_foreign_investment_by_date(from_1m, today, ticker)
        if not f_df.empty:
            col = "보유비율" if "보유비율" in f_df.columns else f_df.columns[-1]
            return f"{float(f_df.iloc[-1][col]):.1f}%"
    except Exception:
        pass
    return "N/A"


def get_korean_stock(ticker: str) -> dict:
    today   = _today()
    from_1m = _date_n_days_ago(30)
    from_1y = _date_n_days_ago(380)

    # 가격/거래량 먼저 (ticker 유효성 확인)
    try:
        ohlcv = pykrx_stock.get_market_ohlcv_by_date(from_1m, today, ticker)
        if ohlcv.empty:
            raise ValueError("empty")
        latest     = ohlcv.iloc[-1]
        prev       = ohlcv.iloc[-2] if len(ohlcv) > 1 else latest
        close      = int(latest["종가"])
        prev_close = int(prev["종가"])
        change_val = close - prev_close
        change_pct = (change_val / prev_close * 100) if prev_close else 0
        vol        = int(latest["거래량"])
        asof       = ohlcv.index[-1].strftime("%Y.%m.%d") + " 종가 기준"
    except Exception:
        return {"error": f"'{ticker}' 종목 데이터를 찾을 수 없습니다."}

    # 나머지 5개 소스 병렬 호출
    with ThreadPoolExecutor(max_workers=5) as ex:
        f_hist    = ex.submit(pykrx_stock.get_market_ohlcv_by_date, from_1y, today, ticker)
        f_info    = ex.submit(_get_name_sector_mktcap, ticker)
        f_naver   = ex.submit(_get_naver_fundamentals, ticker)
        f_yf      = ex.submit(_get_yf_extra, ticker)
        f_income  = ex.submit(lambda: get_dart_financials(ticker) or _get_naver_income(ticker))
        f_foreign = ex.submit(_get_foreign, ticker, from_1m, today)

        try:
            hist_1y = f_hist.result(timeout=15)
            high52  = int(hist_1y["고가"].max())
            low52   = int(hist_1y["저가"].min())
            avgvol  = int(hist_1y["거래량"].tail(20).mean()) if len(hist_1y) >= 20 else vol
        except Exception:
            high52, low52, avgvol = 0, 0, vol

        name, sector, mktcap = f_info.result(timeout=15)
        naver    = f_naver.result(timeout=15)
        yf_extra = f_yf.result(timeout=15)
        income   = f_income.result(timeout=30)
        foreign  = f_foreign.result(timeout=15)

    return {
        "ticker":   ticker,
        "name":     name or ticker,
        "sector":   sector or "N/A",
        "price":    _fmt_krw(close),
        "change":   f"{'+' if change_val >= 0 else ''}{change_val:,}원 ({'+' if change_pct >= 0 else ''}{change_pct:.2f}%)",
        "pos":      change_val >= 0,
        "asof":     asof,
        "per":      naver.get("per", "N/A"),
        "pbr":      naver.get("pbr", "N/A"),
        "eps":      naver.get("eps", "N/A"),
        "mktcap":   mktcap,
        "roe":      yf_extra["roe"],
        "debt":     yf_extra["debt"],
        "evebitda": yf_extra["evebitda"],
        "div":      naver.get("div", "N/A"),
        "beta":     yf_extra["beta"],
        "foreign":  foreign,
        "vol":      str(vol),
        "avgvol":   str(avgvol),
        "high52":   _fmt_krw(high52) if high52 else "N/A",
        "low52":    _fmt_krw(low52)  if low52  else "N/A",
        "income":   income,
    }
